### PR TITLE
feat: Avoid duplicate strings

### PR DESCRIPTION
**Pull request by ynsehoornenborg**
New Analyzer to check duplicate string literal in a class or struct.

### CODE REVIEW

File: Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+		{
+		}
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.EnableConcurrentExecution();
+
+			context.RegisterSyntaxNodeAction(Analyze, SyntaxKind.ClassDeclaration);
+			context.RegisterSyntaxNodeAction(Analyze, SyntaxKind.StructDeclaration);
```
Comment (bcollamore) [inline Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs:32]: Consider an Allowed.DuplicateStrings.txt

File: Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+		}
+
+		public void Analyze(SyntaxNodeAnalysisContext context)
+		{
```
Comment (bcollamore) [inline Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs:36]: Consider analyzing across Project in lieu of File. (Or, config option, idk.)

File: Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
+// © 2023 Koninklijke Philips N.V. See License.md in the project root for license information.
+
+using System.Collections.Generic;
```
Comment (bcollamore) [inline Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs:3]: I would keep this in MaintainabilityAnalyzers. DuplicateCodeAnalyzer is really a special case. (ie., read very slow, etc.) Folks opt into this separate package. Whereas they already have MaintainabilityAnalyzers and should get this one for "free".

File: Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
```
Comment (bcollamore) [inline Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs:6]: On the performance note, this one was relatively slow at 75ms, but still within reason. It's possibly faster to register per Literal and then save in a Dictionary (a la DuplicateCodeAnalyzer). The dictionary can be segmented per file or per project if desired.

File: Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+using Philips.CodeAnalysis.Common;
+
+namespace Philips.CodeAnalysis.DuplicateCodeAnalyzer
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class AvoidDuplicateStringsAnalyzer : SingleDiagnosticAnalyzer
+	{
+		private const string Title = @"Avoid Duplicate Strings";
+		private const string MessageFormat = @"Duplicate string found, first location at line {0}. Consider moving '{1}' into a constant.";
+		private const string Description = @"Duplicate code is less maintainable";
```
Comment (bcollamore) [inline Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs:18]: Idk if strings are code per se. Consider "Duplicate strings are less maintainable."

File: Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+		{
+			var typeDeclarationSyntax = (BaseTypeDeclarationSyntax)context.Node;
+
+			GeneratedCodeDetector detector = new();
+			if (detector.IsGeneratedCode(context))
+			{
+				return;
+			}
+
+			TestHelper testHelper = new();
```
Comment (bcollamore) [inline Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs:45]: Good call...

**Review by bcollamore — DISMISSED** [review]

File: Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs (hunk 2 above)
Comment (ynsehoornenborg) [reply Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs:40]: Done, like in `DuplicateCodeAnalyzer`

File: Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs (hunk 4 above)
Comment (ynsehoornenborg) [reply Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs]: Done

File: Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs (hunk 5 above)
Comment (ynsehoornenborg) [reply Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs]: Corrected

File: Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs (hunk 3 above)
Comment (ynsehoornenborg) [reply Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs]: Moved it

File: Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs (hunk 1 above)
Comment (ynsehoornenborg) [reply Philips.CodeAnalysis.DuplicateCodeAnalyzer/AvoidDuplicateStringsAnalyzer.cs]: For me, if you take the effort to add it to such a file, why don't you fix it by adding it to a `StringConstants.cs` file   instead.

**Review by bcollamore — APPROVED** [review]

**Pull request merged**